onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -2006,10 +2020,29 @@ common::Status TensorrtExecutionProvider::Compile(const std::vector<Node*>& fuse
         }
       }
 
-      // Run TRT inference
-      if (!trt_context->enqueueV2(&buffers[0], stream, nullptr)) {
-        return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "TensorRT EP execution context enqueue failed.");
-      }
+      // Run TRT inference			  
+      if (trt_state->cuda_graph_enable)
+      {
+          if (*cuda_graph == nullptr) {
+            cudaGraph_t graph;
+            *cuda_graph = &(trt_state->cuda_graph_instance);
+            //warm up for cuda graph capturing
+            if (!trt_context->enqueueV2(&buffers[0], stream, nullptr)) {
```

Comment:
Does enqueueV2() synchronize with the GPU before returning ? If not, we may have to wait for the warm-up tasks queued on the stream to finish before the stream capture...

Reply:
why is this warm up even needed?

Reply:
seems like it's for handling a known issue with dynamic shapes? (please add a comment)

Reply:
warm-up is needed to do initialization (flushing any old context) before graph capturing according to Nvidia. CUDA graph still has issue in dynamic shape cases.

Reply:
in the note section of the api doc for enqueuev2 https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_execution_context.html#a2f4429652736e8ef6e19f433400108c7 
seems to allude to dynamic shapes can work if you call enqueuev2() once before graph capture? is it similar to what you are doing?

Reply:
If shape changed, cuda graph needs to be recaptured, which is not desired because the capturing happens in inference.